File: frontend/src/types/invoice.ts
/**
 * Types para sistema de Faturas (Invoices)
 */

export enum InvoiceStatus {
  PENDENTE = 'pendente',
  AGUARDANDO_CONF = 'aguardando_conf',
  PAGO = 'pago',
  ATRASADO = 'atrasado',
  CANCELADO = 'cancelado',
}

export enum MetodoPagamento {
  PIX = 'pix',
  TRANSFERENCIA = 'transferencia',
  BOLETO = 'boleto',
  CARTAO = 'cartao',
  DINHEIRO = 'dinheiro',
  OUTRO = 'outro',
}

export interface Invoice {
  id: string;
  project_id: number;
  numero_fatura: string;
  descricao: string;

  // Valores
  valor: number;
  desconto: number;
  valor_final: number;

  // Datas
  data_emissao: string; // ISO date
  data_vencimento: string; // ISO date
  data_pagamento?: string; // ISO date

  // Status
  status: InvoiceStatus;

  // Documentos
  nota_fiscal_url?: string;
  comprovante_url?: string;
  comprovante_enviado_em?: string; // ISO timestamp

  // Confirmação admin
  confirmado_por?: string;
  confirmado_em?: string; // ISO timestamp
  notas_admin?: string;

  // Método
  metodo_pagamento?: MetodoPagamento;

  // Metadata
  created_at: string;
  updated_at: string;

  // Joins
  project_nome?: string;
  client_nome?: string;

  // Computed fields
  esta_atrasado?: boolean;
  dias_ate_vencimento?: number;
}

export interface InvoiceCreate {
  project_id: number;
  descricao: string;
  valor: number;
  desconto?: number;
  data_vencimento: string; // ISO date
  data_emissao?: string; // ISO date
  metodo_pagamento?: MetodoPagamento;
  notas_admin?: string;
}

export interface InvoiceUpdate {
  descricao?: string;
  valor?: number;
  desconto?: number;
  data_vencimento?: string;
  metodo_pagamento?: MetodoPagamento;
  status?: InvoiceStatus;
  notas_admin?: string;
}

export interface InvoiceStats {
  total_faturado: number;
  total_pendente: number;
  total_atrasado: number;
  total_pago: number;

  quantidade_pendente: number;
  quantidade_atrasado: number;
  quantidade_pago: number;
  quantidade_aguardando_conf: number;

  proximos_vencimentos: Invoice[];
}

export const STATUS_LABELS: Record<InvoiceStatus, string> = {
  [InvoiceStatus.PENDENTE]: 'Pendente',
  [InvoiceStatus.AGUARDANDO_CONF]: 'Aguardando Confirmação',
  [InvoiceStatus.PAGO]: 'Pago',
  [InvoiceStatus.ATRASADO]: 'Atrasado',
  [InvoiceStatus.CANCELADO]: 'Cancelado',
};

export const STATUS_COLORS: Record<InvoiceStatus, string> = {
  [InvoiceStatus.PENDENTE]: 'bg-yellow-100 text-yellow-700',
  [InvoiceStatus.AGUARDANDO_CONF]: 'bg-blue-100 text-blue-700',
  [InvoiceStatus.PAGO]: 'bg-green-100 text-green-700',
  [InvoiceStatus.ATRASADO]: 'bg-red-100 text-red-700',
  [InvoiceStatus.CANCELADO]: 'bg-gray-100 text-gray-700',
};

export const METODO_LABELS: Record<MetodoPagamento, string> = {
  [MetodoPagamento.PIX]: 'PIX',
  [MetodoPagamento.TRANSFERENCIA]: 'Transferência',
  [MetodoPagamento.BOLETO]: 'Boleto',
  [MetodoPagamento.CARTAO]: 'Cartão',
  [MetodoPagamento.DINHEIRO]: 'Dinheiro',
  [MetodoPagamento.OUTRO]: 'Outro',
};

File: backend/app/models/invoice.py
"""
Modelos Pydantic para Faturas (Invoices)
"""
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, computed_field


class InvoiceStatus(str, Enum):
    """Status possíveis de uma fatura"""
    PENDENTE = "pendente"
    AGUARDANDO_CONF = "aguardando_conf"
    PAGO = "pago"
    ATRASADO = "atrasado"
    CANCELADO = "cancelado"


class MetodoPagamento(str, Enum):
    """Métodos de pagamento aceitos"""
    PIX = "pix"
    TRANSFERENCIA = "transferencia"
    BOLETO = "boleto"
    CARTAO = "cartao"
    DINHEIRO = "dinheiro"
    OUTRO = "outro"


class InvoiceBase(BaseModel):
    """Modelo base de fatura"""
    project_id: UUID
    descricao: str = Field(..., min_length=1, max_length=500)
    valor: Decimal = Field(..., gt=0, decimal_places=2)
    desconto: Decimal = Field(default=Decimal("0"), ge=0, decimal_places=2)
    data_vencimento: date
    metodo_pagamento: Optional[MetodoPagamento] = None


class InvoiceCreate(InvoiceBase):
    """Modelo para criação de fatura (usado pelo admin)"""
    data_emissao: Optional[date] = None  # Se não fornecido, usa data atual
    notas_admin: Optional[str] = None


class InvoiceUpdate(BaseModel):
    """Modelo para atualização de fatura"""
    descricao: Optional[str] = Field(None, min_length=1, max_length=500)
    valor: Optional[Decimal] = Field(None, gt=0, decimal_places=2)
    desconto: Optional[Decimal] = Field(None, ge=0, decimal_places=2)
    data_vencimento: Optional[date] = None
    metodo_pagamento: Optional[MetodoPagamento] = None
    status: Optional[InvoiceStatus] = None
    notas_admin: Optional[str] = None


class InvoiceUploadComprovante(BaseModel):
    """Modelo para upload de comprovante pelo cliente"""
    comprovante_url: str = Field(..., min_length=1)


class InvoiceConfirmarPagamento(BaseModel):
    """Modelo para confirmação de pagamento pelo admin"""
    data_pagamento: date
    notas_admin: Optional[str] = None


class Invoice(InvoiceBase):
    """Modelo completo de fatura (retornado pela API)"""
    id: UUID
    numero_fatura: str
    valor_final: Decimal
    data_emissao: date
    data_pagamento: Optional[date] = None
    status: InvoiceStatus

    # URLs de documentos
    nota_fiscal_url: Optional[str] = None
    comprovante_url: Optional[str] = None
    comprovante_enviado_em: Optional[datetime] = None

    # Confirmação admin
    confirmado_por: Optional[UUID] = None
    confirmado_em: Optional[datetime] = None
    notas_admin: Optional[str] = None

    # Metadata
    created_at: datetime
    updated_at: datetime

    # Informações do projeto (join)
    project_nome: Optional[str] = None
    client_nome: Optional[str] = None

    @computed_field
    @property
    def esta_atrasado(self) -> bool:
        """Verifica se a fatura está atrasada"""
        if self.status in [InvoiceStatus.PAGO, InvoiceStatus.CANCELADO]:
            return False
        return self.data_vencimento < date.today()

    @computed_field
    @property
    def dias_ate_vencimento(self) -> int:
        """Dias até o vencimento (negativo se atrasado)"""
        delta = self.data_vencimento - date.today()
        return delta.days

    class Config:
        from_attributes = True


class InvoiceStats(BaseModel):
    """Estatísticas de faturas (para dashboard admin)"""
    total_faturado: Decimal
    total_pendente: Decimal
    total_atrasado: Decimal
    total_pago: Decimal

    quantidade_pendente: int
    quantidade_atrasado: int
    quantidade_pago: int
    quantidade_aguardando_conf: int

    proximos_vencimentos: list[Invoice] = Field(default_factory=list)
